searx/doc.py

```diff
@@ -0,0 +1,290 @@
+import mistletoe
+
+from . import get_setting
+from .version import GIT_URL
+
+
+def _usage(render_types):
+    doc = OnlineDoc()
+    msg = """python -m %(name)s [-h] <pagename> [<render by>]
+
+pagename: [%(pagename)s]
+render by: [%(render_by)s] default is *Offline*
+"""
+    ctx = {
+        'name': 'searx.doc',
+        'pagename': '|'.join(doc.toc),
+        'render_by': '|'.join(render_types.keys()),
+    }
+    return msg % ctx
+
+
+def main():
+    """Command line to print a page of the user documentation to stdout"""
+    _, pagename, render_by = (sys.argv + ['', '', ''])[:3]
+    render_by = render_by or 'Offline'
+
+    render_types = {'Markup': Markup, 'Markdown': Markdown, 'Offline': Offline, 'Mistletoe': Mistletoe}
+    if '-h' in sys.argv:
+        print(_usage(render_types), file=sys.stderr)
+        return 42
+
+    markup_cls = render_types.get(render_by)
+    if markup_cls is None:
+        print(_usage(render_types), file=sys.stderr)
+        print("ERROR: <render by> %s is unknown" % render_by, file=sys.stderr)
+        return 1
+
+    doc = OnlineDoc(markup_cls())
+    if pagename is None or markup_cls is None or pagename not in doc.toc:
+        print(_usage(render_types), file=sys.stderr)
+        if pagename is None:
+            print("ERROR: missing <pagename>", file=sys.stderr)
+        elif pagename not in doc.toc:
+            print("ERROR: unknown <pagename> '%s'" % pagename, file=sys.stderr)
+        return 1
+
+    # render online doc
+    # -----------------
+
+    # To render the online documentation a Flask application context is needed.
+    # The url_for functions in the jinja templates need all routes to be
+    # registered in the Flask app.
+
+    # pylint: disable=import-outside-toplevel
+
+    from searx import settings
+
+    settings['server']['secret_key'] = "x"
+    from searx.webapp import app
+
+    # Specify base_url so that url_for() works for base_urls that have a
+    # non-root path // If base_url is specified, then these values from are
+    # given preference over any Flask's generics (see flaskfix.py).
+
+    base_url = get_setting('server.base_url', None)
+    with app.test_request_context(base_url=base_url):
+        page = doc.render(pagename)
+
+    if isinstance(page, HTMLPage):
+        page = page.html
+
+    ret_val = 0
+    if page is None:
+        print("ERROR: can't render '%s' '%s'" % (pagename, render_by), file=sys.stderr)
+        ret_val = 42
+    else:
+        print(page)
+
+    # The searx.webapp import from above fires some HTTP requests, thats
+    # why we get a RuntimeError::
+    #
+    #     RuntimeError: The connection pool was closed while 1 HTTP \
+    #       requests/responses were still in-flight.
+    #
+    # Closing network won't help ..
+    #   from searx.network import network
+    #   network.done()
+
+    # waiting some seconds before ending the comand line was the only solution I
+    # found ..
+    import time
+
+    time.sleep(3)
```

Comment:
> So do we load the application?

You asking "why do we load the app"? ..  because the `url_for` command needs the  flask context. See the comment above https://github.com/searxng/searxng/pull/930/files#diff-b439f9f6bc445fcbb1be1c7d0314a845126d56ba235e52361c9918d943d5733aR84-R90


```
    # To render the online documentation a Flask application context is needed.
    # The url_for functions in the jinja templates need all routes to be
    # registered in the Flask app.
```

Reply:
Call sequence:
* `pyenv.cmd python -m searx.doc search-syntax Offline`
* Pick the [Offline](https://github.com/searxng/searxng/blob/5d45cd6eb18095e0bb3e5715c58072e68c4aa727/searx/doc.py#L64)
* [`main(): doc = OnlineDoc(Offline())`](https://github.com/searxng/searxng/blob/5d45cd6eb18095e0bb3e5715c58072e68c4aa727/searx/doc.py#L75)
* [main():  Initialize `searx.webapp`](https://github.com/searxng/searxng/blob/5d45cd6eb18095e0bb3e5715c58072e68c4aa727/searx/doc.py#L93-L96)
* [`main(): page = doc.render(pagename)`](https://github.com/searxng/searxng/blob/5d45cd6eb18095e0bb3e5715c58072e68c4aa727/searx/doc.py#L102-L104)
* [`OnlineDoc.render(): page = self.markup.render(content)`](https://github.com/searxng/searxng/blob/5d45cd6eb18095e0bb3e5715c58072e68c4aa727/searx/doc.py#L283)
* [`Markup.render`](https://github.com/searxng/searxng/blob/5d45cd6eb18095e0bb3e5715c58072e68c4aa727/searx/doc.py#L153-L157)
* [`Offline.get_ctx`](https://github.com/searxng/searxng/blob/5d45cd6eb18095e0bb3e5715c58072e68c4aa727/searx/doc.py#L228-L235) : here we don't call `url_for`

Reply:
> so `url_for` won't be called, so there is no need for the app to be initialized.

Not for the `url_for` (`link` & `search`) function but may be in the future we want to use other information or functions from the Flask context? .. if we will never use any information from the Flask context in the project documentation (aka Offline) .. we can simplify the code a lot (like you have done in your branch).

What do you think, should I simplify the code of this PR?

Reply:
> should I simplify the code

Yes!

Currently, I feel `searx/doc.py` is a tool that should be in `searxng_extra`.
IMO, for the runtime, there are too many abstractions.

Reply:
Hm .. just I have another thought (sorry) .. I think the Flask context is also useful when the project documentation is build.  Assume the `base_url` is set:

https://github.com/searxng/searxng/blob/927aa71133839ce5713b742b43285953301adf36/searx/settings.yml#L45

then we can also compile links in the the project documentation / that is linked by

https://github.com/searxng/searxng/blob/927aa71133839ce5713b742b43285953301adf36/searx/settings.yml#L9

Since we do not have a base_url in our project, we do not need this, but forks who build their own online documentation will find it useful I think.

With said in mind, what do you think about?

Reply:
I'm not sure to understand.

Do you suggest that some forks may wish to create links from the Sphinx documentation to their own instance?
So in this case `url_for` would be useful for them.

I can't say, but I have seen requests in that way and so I would say let's skip this for now.

Reply:
> Do you suggest that some forks may wish to create links from the Sphinx documentation to their own instance?

yes.